File: src/infrastructure/table.ts
export function formatTable(
  rows: Array<Record<string, string | number | null | undefined>>,
  columns: string[],
): string {
  if (rows.length === 0) {
    return columns.join("  ") + "\n" + columns.map(() => "--").join("  ");
  }

  const stringRows = rows.map((row) =>
    columns.map((col) => {
      const value = row[col];
      if (value === null || value === undefined) return "";
      return String(value);
    }),
  );

  const widths = columns.map((col, i) =>
    Math.max(
      col.length,
      ...stringRows.map((r) => r[i]?.length ?? 0),
      2,
    ),
  );

  const pad = (cells: string[]) =>
    cells.map((cell, i) => cell.padEnd(widths[i]!)).join("  ");

  const header = pad(columns);
  const sep = widths.map((w) => "-".repeat(w)).join("  ");
  const body = stringRows.map((r) => pad(r)).join("\n");
  return `${header}\n${sep}\n${body}`;
}

File: src/infra/table.rs
pub fn format_table(rows: &[Vec<String>], columns: &[&str]) -> String {
    if rows.is_empty() {
        let header = columns.join("  ");
        let sep = columns.iter().map(|_| "--").collect::<Vec<_>>().join("  ");
        return format!("{header}\n{sep}");
    }
    let mut widths: Vec<usize> = columns.iter().map(|c| c.len().max(2)).collect();
    for row in rows {
        for (i, cell) in row.iter().enumerate() {
            if i < widths.len() {
                widths[i] = widths[i].max(cell.len());
            }
        }
    }
    let pad = |cells: &[String]| {
        cells
            .iter()
            .enumerate()
            .map(|(i, cell)| {
                let w = widths.get(i).copied().unwrap_or(cell.len());
                format!("{cell:<w$}")
            })
            .collect::<Vec<_>>()
            .join("  ")
    };
    let header = pad(
        &columns
            .iter()
            .map(|c| c.to_string())
            .collect::<Vec<_>>(),
    );
    let sep = widths
        .iter()
        .map(|w| "-".repeat(*w))
        .collect::<Vec<_>>()
        .join("  ");
    let body = rows
        .iter()
        .map(|r| pad(r))
        .collect::<Vec<_>>()
        .join("\n");
    format!("{header}\n{sep}\n{body}")
}
